### PR TITLE
Cache dependencies on github actions (#1296)

### LINKED ISSUE #1296: Enhance github action speed by caching dependencies
**Description**
Per [github](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows), we can try to cache the dependencies to increase the speed of each execution. 

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Resolve #1296

Caching for both npm and pipenv (for tests only). Deployment/build will still do everything from scratch.
Caching was also added for frontend build.

For npm, the dependency was declared and those will be used to hash as the cache name. 

For pipenv, the recommended file is `Pipfile.lock`. But since we don't use `Pipfile.lock` (mainly due to multiple os support), `Pipfile` will be used. We might have to manually delete the cache from [here](https://github.com/Avaiga/taipy/actions/caches) if any dev packages get a major/breaking update. Pipenv caching will work both for the min Pipfile and max Pipfile as those will be produce 2 different caches due to the different file hashes.

The cache will be removed automatically after 7 days if not used in any action. There is no limit on the number of cache, but there is a hard limit of 10 GB of cache per repo.

### CODE REVIEW

**Review by FredLL-Avaiga — DISMISSED** [review]
I'm learning everyday 👍

**Review by FabienLelaquais — DISMISSED** [review]
So that was impressive! 🤯

File: .github/workflows/partial-tests.yml [hunk 1]
```diff
@@ -64,9 +66,17 @@ jobs:
         if: matrix.os == 'macos-13'
         run: brew install libmagic
 
-      - uses: actions/setup-node@v4
+      - name: Setup Node
+        if: steps.changes.outputs.gui == 'true' || steps.changes.outputs.gui-core == 'true'
```
Comment (jrobinAV) [inline .github/workflows/partial-tests.yml:70]: 👍🏻

File: .github/workflows/partial-tests.yml [hunk 2]
```diff
@@ -128,7 +138,17 @@ jobs:
...
+        with:
+          python-version: ${{matrix.python-version}}
+          cache: 'pipenv'
+          cache-dependency-path: '**/Pipfile'
+
+      - name: Setup Python without cache
+        if: steps.changes.outputs.core != 'true'
+        uses: actions/setup-python@v5
         with:
           python-version: ${{matrix.python-version}}
```
Comment (jrobinAV) [inline .github/workflows/partial-tests.yml]: The `submit_tests` job is only run if some core files are touched. This step is run when core files are not touched. I don't get its purpose.

File: .github/workflows/partial-tests.yml [hunk 3]
```diff
@@ -167,7 +187,17 @@ jobs:
...
```
Comment (jrobinAV) [inline .github/workflows/partial-tests.yml]: same here.

File: .github/workflows/partial-tests.yml [hunk 4]
```diff
@@ -128,7 +138,17 @@ jobs:
             core:
               - 'taipy/core/**'
 
-      - uses: actions/setup-python@v5
+      - name: Setup Python with cache
+        if: steps.changes.outputs.core == 'true'
```
Comment (jrobinAV) [inline .github/workflows/partial-tests.yml:157]: Good catch. the filter was missing.

**Review by jrobinAV — COMMENTED** [review]
This PR only covers partial tests. Why not the overall tests as well?

**Review by FabienLelaquais — DISMISSED** [review]
🤯🤯🤯

**Comment by dinhlongviolin1** [comment]
> This PR only covers partial tests. Why not the overall tests as well?

The overall tests uses the install/action.yml file for installation of the dependencies. The cache is already implemented in that file.

File: .github/workflows/partial-tests.yml (hunk 4 above)
Comment (dinhlongviolin1) [reply .github/workflows/partial-tests.yml:157]: Thanks!

File: .github/workflows/partial-tests.yml (hunk 3 above)
Comment (dinhlongviolin1) [reply .github/workflows/partial-tests.yml]: I will remove it. Thanks!

**Comment by dinhlongviolin1** [comment]
> This PR only covers partial tests. Why not the overall tests as well?

We can also remove the python setup within overall-tests.yml file since it will be setup in the install/action.yml file. @florian-vuillemot can you confirm this

**Review by jrobinAV — DISMISSED** [review]

**Review by FabienLelaquais — DISMISSED** [review]
👍👍

**Review by florian-vuillemot — DISMISSED** [review]

**Comment by dinhlongviolin1** [comment]
Caching the frontend build is also added!

**Review by FredLL-Avaiga — DISMISSED** [review]

File: tools/frontend/hash_source.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    combined_hash = combined_hasher.hexdigest()
+    return file_hashes, combined_hash
+
+
+if __name__ == "__main__":
+    frontend_folder = "frontend"
+    file_hashes, combined_hash = hash_files_in_frontend_folder(frontend_folder)
+    # for path, file_hash in sorted(file_hashes.items()):
+    #     print(f"{path}: {file_hash}")
```
Comment (jrobinAV) [inline tools/frontend/hash_source.py:71]: To remove

**Review by jrobinAV — APPROVED** [review]

File: tools/frontend/hash_source.py (hunk 5 above)
Comment (dinhlongviolin1) [reply tools/frontend/hash_source.py:71]: I will keep that one for local debugging when needed. Thanks for spotting that JR!

File: tools/frontend/hash_source.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            if any(ignore in root for ignore in ignore_folder):
+                continue
+            # Sort before looping to ensure consistent cache key
+            for file in sorted(files):
+                for ignore in ignore_file_name:
+                    if ignore in file:
+                        continue
+                for ignore in ignore_file_extension:
+                    if file.endswith(ignore):
+                        continue
```
Comment (florian-vuillemot) [inline tools/frontend/hash_source.py]: What is the purpose of these two conditions? If these conditions are met, the `continue` is applied to the first `for` and not to both. Consequently, they do nothing. Do you agree?

Comment (jrobinAV) [reply tools/frontend/hash_source.py]: Indeed

**Review by jrobinAV — DISMISSED** [review]

**Comment by joaoandre-avaiga** [comment]
LGTM! Great job @dinhlongviolin1

**Review by joaoandre-avaiga — DISMISSED** [review]

**Comment by jrobinAV** [comment]
@dinhlongviolin1 Let's merge it if that's fine for everyone so we can finally benefit from it.

File: tools/frontend/hash_source.py (hunk 6 above)
Comment (dinhlongviolin1) [reply tools/frontend/hash_source.py]: Thanks for the finding @florian-vuillemot. I will refactor it

**Review by florian-vuillemot — APPROVED** [review]

**Review by FredLL-Avaiga — APPROVED** [review]

**Review by jrobinAV — APPROVED** [review]

**Review by florian-vuillemot — APPROVED** [review]

**Review by joaoandre-avaiga — APPROVED** [review]

**Pull request merged**